app/models/user.rb

```diff
@@ -1,32 +1,74 @@
 class User < ApplicationRecord
-  attr_accessor :remember_token
+  attr_accessor :remember_token, :activation_token, :reset_token
+  before_save   :downcase_email
+  before_create :create_activation_digest
   before_save{email.downcase!}
   validates :name,  presence: true, length: {maximum: Settings.size.max_name}
   VALID_EMAIL_REGEX = /\A[\w+\-.]+@[a-z\d\-.]+(\.[a-z\d\-]+)*\.[a-z]+\z/i
   validates :email, presence: true, length: {maximum: Settings.size.max_email}, format: {with: VALID_EMAIL_REGEX}, uniqueness: {case_sensitive: false}
   has_secure_password
   validates :password, presence: true, length: {minimum: Settings.size.min_pass}, allow_nil: true
 
-  def self.digest string
-    cost = ActiveModel::SecurePassword.min_cost ? BCrypt::Engine::MIN_COST : BCrypt::Engine.cost
-    BCrypt::Password.create(string, cost: cost)
-  end
+  class << self
+    def digest string
+      cost = if ActiveModel::SecurePassword.min_cost
+               BCrypt::Engine::MIN_COST
+             else
+               BCrypt::Engine.cost
+             end
+      BCrypt::Password.create string, cost: cost
+    end
 
-  def self.new_token
-    SecureRandom.urlsafe_base64
+    def new_token
+      SecureRandom.urlsafe_base64
+    end
   end
 
   def remember
     self.remember_token = User.new_token
-    update_attribute(:remember_digest, User.digest(remember_token))
+    update_attribute :remember_digest, User.digest(remember_token)
   end
 
-  def authenticated? remember_token
-    return false if remember_digest.nil?
-    BCrypt::Password.new(remember_digest).is_password?(remember_token)
+  def authenticated? attribute, token
+    digest = send "#{attribute}_digest"
```

Comment:
`send` để làm gì e ?

Reply:
chưa trả lời này e

> `send` để làm gì e ?

Reply:
.

Reply:
> `send` để làm gì e ?

dạ dùng để gửi thông báo ạ

Reply:
gửi thông báo gì thế e ?

Reply:
> gửi thông báo gì thế e ?

theo em tìm hiểu được thì nó gửi thông báo mail cho người dùng. và BCrypt để lưu trữ mật khẩu an toàn cho người dùng!

Reply:
sai nhé, e tìm hiểu lại về method `send` đi e

Reply:
> sai nhé, e tìm hiểu lại về method `send` đi e

Phương thức send cho phép chúng ta gọi một phương thức tới một đối tượng cụ thể mà chúng ta chưa xác định được chính xác tên của phương thức đó cho đến khi nó được thực thi.  
là cái này đúng ko ạ? nãi h em tưởng nó cũng gưi để bycrypt password :(